extension/observer/endpointswatcher.go

```diff
@@ -15,89 +15,153 @@
+// to poll for endpoints by embedding this struct and using NewEndpointsWatcher().
 type EndpointsWatcher struct {
-	Endpointslister   EndpointsLister
-	RefreshInterval   time.Duration
-	existingEndpoints map[EndpointID]Endpoint
+	EndpointsLister EndpointsLister
+	RefreshInterval time.Duration
+
+	// subscribed Notify instances ~sync.Map(map[NotifyID]Notify)
+	toNotify sync.Map
+	// map of NotifyID to known endpoints for that Notify (subscriptions can occur at different times in service startup).
+	// ~sync.Map(map[NotifyID]map[EndpointID]Endpoint)
+	existingEndpoints sync.Map
 	stop              chan struct{}
+	once              *sync.Once
+	logger            *zap.Logger
 }
 
-// ListAndWatch runs ListEndpoints on a regular interval and keeps the list.
-func (ew *EndpointsWatcher) ListAndWatch(listener Notify) {
-	ew.existingEndpoints = map[EndpointID]Endpoint{}
-	ew.stop = make(chan struct{})
+func NewEndpointsWatcher(endpointsLister EndpointsLister, refreshInterval time.Duration, logger *zap.Logger) *EndpointsWatcher {
+	return &EndpointsWatcher{
+		EndpointsLister:   endpointsLister,
+		RefreshInterval:   refreshInterval,
+		existingEndpoints: sync.Map{},
+		toNotify:          sync.Map{},
+		stop:              make(chan struct{}),
+		once:              &sync.Once{},
+		logger:            logger,
+	}
+}
 
-	// Do the initial listing immediately so that services can be monitored ASAP.
-	ew.refreshEndpoints(listener)
+// ListAndWatch runs EndpointsLister.ListEndpoints() on a regular interval and keeps track of the results
+// for alerting subscribed Notify's of the entries
+func (ew *EndpointsWatcher) ListAndWatch(notify Notify) {
+	ew.once.Do(func() {
+		go func() {
+			ticker := time.NewTicker(ew.RefreshInterval)
+			defer ticker.Stop()
+
+			for {
+				select {
+				case <-ew.stop:
+					return
+				case <-ticker.C:
+					var toNotify []NotifyID
+					ew.toNotify.Range(func(notifyID, _ interface{}) bool {
+						toNotify = append(toNotify, notifyID.(NotifyID))
+						return true
+					})
+					ew.notifyOfEndpoints(toNotify...)
+				}
+			}
+		}()
+	})
 
-	go func() {
-		ticker := time.NewTicker(ew.RefreshInterval)
-		defer ticker.Stop()
+	ew.toNotify.Store(notify.ID(), notify)
+	ew.notifyOfEndpoints(notify.ID())
+}
 
-		for {
-			select {
-			case <-ew.stop:
-				return
-			case <-ticker.C:
-				ew.refreshEndpoints(listener)
-			}
-		}
-	}()
+func (ew *EndpointsWatcher) Unsubscribe(notify Notify) {
+	ew.toNotify.Delete(notify.ID())
+	ew.existingEndpoints.Delete(notify.ID())
 }
 
-// refreshEndpoints updates the listener with the latest list
-// of active endpoints.
-func (ew *EndpointsWatcher) refreshEndpoints(listener Notify) {
-	latestEndpoints := ew.Endpointslister.ListEndpoints()
+// notifyOfEndpoints alerts subscribed Notify instances by their NotifyID of Endpoint events,
+// updating their internal store of last Endpoints advertised.
+func (ew *EndpointsWatcher) notifyOfEndpoints(notifyIDs ...NotifyID) {
```

Comment:
This function is pretty large and was tough to comprehend.  Is there a good way to break any functionality out into helper functions?

Reply:
Yeah, I didn't really change the existing implementation on the first go but took a crack at separating out some helpers (and added concurrency to the main loop).